libbpf-tools/klockstat.bpf.c

```diff
@@ -0,0 +1,230 @@
+	__type(key, struct task_lock);
+	__type(value, struct lockholder_info);
+} lockholder_map SEC(".maps");
+
+/*
+ * Keyed by stack_id.
+ *
+ * Multiple call sites may have the same underlying lock, but we only know the
+ * stats for a particular stack frame.  Multiple tasks may have the same
+ * stackframe.
+ */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, s32);
+	__type(value, struct lock_stat);
+} stat_map SEC(".maps");
+
+static bool tracing_task(u64 task_id)
+{
+	u32 tgid = task_id >> 32;
+	u32 pid = task_id;
+
+	if (targ_tgid && targ_tgid != tgid)
+		return false;
+	if (targ_pid && targ_pid != pid)
+		return false;
+	return true;
+}
+
+static void lock_contended(void *ctx, struct mutex *lock)
+{
+	u64 task_id;
+	struct lockholder_info li[1] = {0};
+	struct task_lock tl = {};
+
+	if (targ_lock && targ_lock != lock)
+		return;
+	task_id = bpf_get_current_pid_tgid();
+	if (!tracing_task(task_id))
+		return;
+
+	li->task_id = task_id;
+	/*
+	 * Skip 4 frames, e.g.:
+	 *       __this_module+0x34ef
+	 *       __this_module+0x34ef
+	 *       __this_module+0x8c44
+	 *             mutex_lock+0x5
+	 *
+	 * Note: if you make major changes to this bpf program, double check
+	 * that you aren't skipping too many frames.
+	 */
+	li->stack_id = bpf_get_stackid(ctx, &stack_map,
+				       4 | BPF_F_FAST_STACK_CMP);
+	/* Legit failures include EEXIST */
+	if (li->stack_id < 0)
+		return;
+	li->try_at = bpf_ktime_get_ns();
+
+	tl.task_id = task_id;
+	tl.lock_ptr = (u64)lock;
+	bpf_map_update_elem(&lockholder_map, &tl, li, BPF_ANY);
+}
+
+static void lock_acquired(struct mutex *lock)
+{
+	u64 task_id;
+	struct lockholder_info *li;
+	struct task_lock tl = {};
+
+	if (targ_lock && targ_lock != lock)
+		return;
+	task_id = bpf_get_current_pid_tgid();
+	if (!tracing_task(task_id))
+		return;
+
+	tl.task_id = task_id;
+	tl.lock_ptr = (u64)lock;
+	li = bpf_map_lookup_elem(&lockholder_map, &tl);
+	if (!li)
+		return;
+
+	li->acq_at = bpf_ktime_get_ns();
+}
+
+static void account(struct lockholder_info *li)
+{
+	struct lock_stat *ls;
+	u64 delta;
+
+	/*
+	 * Multiple threads may have the same stack_id.  Even though we are
+	 * holding the lock, dynamically allocated mutexes can have the same
+	 * callgraph but represent different locks.  They will be accounted as
+	 * the same lock, which is what we want, but we need to use atomics to
+	 * avoid corruption, especially for the total_time variables.
+	 */
+	ls = bpf_map_lookup_elem(&stat_map, &li->stack_id);
+	if (!ls) {
+		struct lock_stat fresh = {0};
+
+		bpf_map_update_elem(&stat_map, &li->stack_id, &fresh, BPF_ANY);
+		ls = bpf_map_lookup_elem(&stat_map, &li->stack_id);
+		if (!ls)
+			return;
+	}
+
+	delta = li->acq_at - li->try_at;
+	__sync_fetch_and_add(&ls->acq_count, 1);
+	__sync_fetch_and_add(&ls->acq_total_time, delta);
+	if (delta > READ_ONCE(ls->acq_max_time)) {
+		WRITE_ONCE(ls->acq_max_time, delta);
+		WRITE_ONCE(ls->acq_max_id, li->task_id);
```

Comment:
Why do you need these `READ_ONCE` and `WRITE_ONCE`s ?

Reply:
in part, for documentation that those are read from and written to concurrently.  that's what i'd do for in-kernel code.  

the bpf compiler might make those u64 accesses atomic, but i don't know if that is a guarantee, and i'd rather not have issues with load/store tearing.